Components: Make Toolbar accessible

diegohaz opened the pull request:
## Description
Fixes #15331.
Fixes #3383.
Related to #16514.

<!-- Please describe what you have changed or added -->

This PR fixes `Toolbar` transforming it into an **actual toolbar**, addressing the accessibility concerns discussed on #15331, and leveraging the design discussions on #16514.

I tried to touch the API as little as possible. But we have a few changes, all backward compatible:

1. Current `Toolbar` has been renamed to `ToolbarGroup`.
2. New `Toolbar` needs an `accessibilityLabel` prop.
3. New `Toolbar` requires that its buttons are rendered using `@wordpress/components`.

### 1. Current `Toolbar` has been renamed to `ToolbarGroup`

The current `Toolbar` on master still works, but now it will emit a deprecation warning with a suggestion to use `ToolbarGroup` instead, which name better describes its semantics.

`ToolbarGroup` has the same implementation as the current `Toolbar` on master. In fact, most of the code was copied and pasted to the new location.

```jsx
import { ToolbarGroup } from '@wordpress/components';

<ToolbarGroup>
  ...
</ToolbarGroup>

// or
<ToolbarGroup controls={...} />

// or
<ToolbarGroup isCollapsed controls={...} />
```

Just like today, `<ToolbarGroup isCollapsed />` only accepts `controls`, not `children`.

### 2. New `Toolbar` needs an `accessibilityLabel` prop.

To use the new `Toolbar`, it's necessary to pass an `accessibilityLabel` prop to the component. Other than that, the API remains the same:
```jsx
import { Toolbar, ToolbarButton, ToolbarGroup } from '@wordpress/components';

<Toolbar accessibilityLabel="Block options">
  <ToolbarButton icon="thumbs-up" title="Thumbs Up" onClick={...} />
  <ToolbarButton icon="thumbs-down" title="Thumbs Down" onClick={...} />
  <ToolbarGroup>
    <ToolbarButton ... />
    <ToolbarButton ... />
  </ToolbarGroup>
   ...
</Toolbar>
```

When using `Toolbar` without `accessibilityLabel`, the component will fallback to `ToolbarGroup` and emit a warning.

### 3. New `Toolbar` requires that its buttons are rendered using `@wordpress/components`.

Like `Button`, `IconButton`, `ToolbarButton` etc.

Ideally, only `ToolbarButton` would be accepted, and it would be flexible enough to render all kinds of toolbar items. Unfortunately, that's not the case for `ToolbarButton`, and even Gutenberg uses different components as toolbar items to achieve different results.

This means that block authors that aren't using `@wordpress/components` for toolbar items will see a warning.
```jsx
import {
  BlockControls,
  Button,
  IconButton,
  ToolbarButton,
  DropdownMenu,
} from '@wordpress/components';

// still works, but will fallback to the old navigable menu
// less accessible, but without breaking changes
<BlockControls>
  <button />
  <button />
  <button />
  <button />
</BlockControls>

// after
<BlockControls>
  <Button />
  <IconButton />
  <ToolbarButton />
  <DropdownMenu />
</BlockControls>
```

### Other stuff

#### How is this different from `NavigableToolbar`

`NavigableToolbar` finds all the tabbable elements within the toolbar whenever the user presses left and arrow keys. It determines which one is active and focuses the previous/next one. To conform with the [WAI-ARIA recommendations](https://www.w3.org/TR/wai-aria-1.1/#toolbar), it still needs to:
- set `tabIndex="-1"` on inactive items so the toolbar has only one tab stop.
- keep record of the last active item so focus will move back to it the user tabs into the toolbar.
- listen to <kbd>PageUp</kbd> <kbd>PageDown</kbd> <kbd>Home</kbd> <kbd>End</kbd> keys.

The new `Toolbar` solves all these issues. Also, it registers `Button`s when they're mounted and unregisters them when they're unmounted. Because of that, it doesn't need to run a query on the DOM every time an arrow key is pressed, which leads to a better performance.

#### Reakit

This PR introduces a new dependency: [Reakit](https://github.com/reakit/reakit). It's an open source library that provides low level components and hooks to address common accessibility aspects of design systems. It has [1000s of lines of tests](https://github.com/reakit/reakit/blob/b8ff9c2222dc20194983e5ad58ece7bef88012a0/packages/reakit/src/Menu/__tests__/index-test.tsx) to make sure accessible behaviors work. It's being used by projects like [CodeSandbox](https://github.com/reakit/reakit/issues/434#issuecomment-532116655) and [Gatsby](https://github.com/system-ui/theme-ui/blob/04dcb13aa291366b0292b3f76978d80d65e00d7b/packages/editor/src/ColorPicker.js#L10-L15). And is used here to add the [roving tabindex](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex) method to `Toolbar`.

`reakit/Toolbar` adds around `3.9 kB` gzipped to the bundle. My plan is to address higher-level cases like `Toolbar`, `Popover` or whatever is more urgent first. I want to see how well Reakit fits into the current `@wordpress/components` implementation while trying to touch the components' API as little as possible.

Once we have a few high-level components working with proper accessibility and less dependent of `@wordpress/components`' abstractions, we can go deeper and determine what to do with things like `NavigableContainer`.

I think that's the path that will give us more flexibility to change things (even removing Reakit later, if needed) at the same time we provide more accessible ready-to-use components faster.

### Next steps

As I said, I tried to touch the `Toolbar` API as little as possible. But this should be improved. `ToolbarButton` must be more flexible and work for more cases besides only `IconButton`. So, next steps would be improving `Toolbar` API. Some discussions are happening on #16514.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Checked the header toolbar and the block toolbar of many (if not all) blocks. Many end-to-end tests failed on the first iterations, so I went through each of them and fixed the issues. 

## Screenshots <!-- if applicable -->

<details><summary>Storybook</summary>

![Oct-26-2019 22-52-06](https://user-images.githubusercontent.com/3068563/67629920-e6c9e880-f843-11e9-8332-f96bab3bc5f1.gif)

</details>

<details><summary>Playground</summary>

![Oct-26-2019 23-53-13](https://user-images.githubusercontent.com/3068563/67630339-d0c02600-f84b-11e9-9f3d-74e5a7d64c77.gif)

</details>

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

diegohaz committed:
Add initial accessible toolbar implementation

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Fix ToolbarButton style by putting box-sizing on the component styles

diegohaz committed:
Support current ToolbarButton usage

diegohaz committed:
Remove useMemo

This should be done upstream

diegohaz committed:
Re-export current native versions

diegohaz committed:
Fix AccessibleToolbarButton overwriting existing IconButton props

diegohaz committed:
Add label to Toolbar

diegohaz committed:
Create stories

diegohaz committed:
Upgrade reakit dependency

diegohaz committed:
Fix ToolbarContext value

diegohaz committed:
Update package-lock.json

diegohaz committed:
Use accessibilityLabel instead of label

diegohaz committed:
Merge toggleProps event handlers in DropdownMenu

diegohaz committed:
Add ToolbarGroup component

diegohaz committed:
Use ToolbarGroup on Toolbar legacy

diegohaz committed:
Update Toolbar stories

diegohaz committed:
Update Toolbar README.md

diegohaz committed:
Fix ToolbarGroup not being included in the tab order on the legacy usage

jsnajdr commented on packages/components/src/toolbar-button/accessible-toolbar-button.js:
```diff
@@ -0,0 +1,23 @@
+/**
+ * External dependencies
+ */
+import { useToolbarItem } from 'reakit/Toolbar';
```
What does `useToolbarItem` do?

diegohaz replied:
It’s the same as `<ToolbarItem>`, except that it doesn’t return a component. It only returns the HTML props that could be passed to a native `<button>`, for example. That’s how Reakit composes multiple “components” without polluting the React component tree.

More info:
- https://reakit.io/docs/basic-concepts/#props-hooks
- https://reakit.io/docs/composition/#props-hooks

diegohaz replied:
More specifically, it adds a new item to the toolbar state (so it knows what’s the previous and next items) and handles key down events (by returning an `onKeyDown` prop) so pressing arrow keys will move focus between items.

diegohaz replied:
Yes! Reakit can be used entirely without importing any actual React component. Example: https://codesandbox.io/s/reakit-toolbar-props-hooks-1bxp2

diegohaz committed:
Update AccessibleToolbarButton

diegohaz committed:
Refactor ToolbarContext and deprecate Toolbar being used as collapsible group

diegohaz committed:
Update stories

diegohaz committed:
Refactor ToolbarGroup

gziolo commented:
It's great to see how it shapes up so far. I have some questions which we will need to answer as this PR progresses.

> Besides that, this PR introduces a new dependency: [Reakit](https://github.com/reakit/reakit). It's an open source library that provides low level components and hooks to address common accessibility aspects of design systems. It has [1000s of lines of tests](https://github.com/reakit/reakit/blob/b8ff9c2222dc20194983e5ad58ece7bef88012a0/packages/reakit/src/Menu/__tests__/index-test.tsx) to make sure accessible behaviors work. It's being used by projects like [CodeSandbox](https://github.com/reakit/reakit/issues/434#issuecomment-532116655) and [Gatsby](https://github.com/system-ui/theme-ui/blob/04dcb13aa291366b0292b3f76978d80d65e00d7b/packages/editor/src/ColorPicker.js#L10-L15). And is used here to add the [roving tabindex](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex) method to Toolbar.

According to https://bundlephobia.com/result?p=reakit@1.0.0-beta.9:

![Screen Shot 2019-10-17 at 10 41 31](https://user-images.githubusercontent.com/699132/66992782-b3ac8a00-f0ca-11e9-812c-d9325095ed71.png)

What's the actual impact in this PR. I bet it's going to benefit from tree-shaking but it still might have some impact.

Can you expand on how do you plan to address the duplication of similar solutions that already exist in Gutenberg for other components? I'm thinking about alternative approaches for roving, click outside detection, popovers, modals, etc. There is an existing set of Higher-Order components and small utilities that are coded differently in Reakit.

diegohaz commented:
@gziolo 

> What's the actual impact in this PR. I bet it's going to benefit from tree-shaking but it still might have some impact.

On a simple `create-react-app` project, `reakit/Toolbar` and its dependencies add a few less than `4kB` to the bundle (after gzip):

![image](https://user-images.githubusercontent.com/3068563/66997361-87274c80-f08f-11e9-8bbd-36a4dccde5c3.png)

If more modules are used, the added size would likely be smaller as they might share the same dependencies (`reakit/Rover`, for example).

> Can you expand on how do you plan to address the duplication of similar solutions that already exist in Gutenberg for other components?

My plan is to address higher-level cases like Toolbar, Popover or whatever is more urgent first (and I'd appreciate suggestions on next components 😁). I want to see how well Reakit fits into the current `@wordpress/components` implementation while trying to touch the components' API as little as possible.

This may include replacing `NavigableContainer` by, for example, `reakit/Rover` in some components, but not all of them. There will be some duplication for some time. But, once we have a few high-level components working with proper accessibility and less dependent of `@wordpress/components`' abstractions, we can go deeper and determine what to do with things like `NavigableContainer`.

I think that's the path that will give us more flexibility to change things (even removing Reakit later, if needed) at the same time we provide more accessible ready-to-use components faster. But, of course, I'm always open to new ideas. :)

diegohaz committed:
Export ToolbarGroup

diegohaz committed:
Fix tests

diegohaz committed:
Fix Styling

diegohaz commented:
I added a deprecation warning to `Toolbar` and now many end-to-end tests are breaking because of the `console.warn`:

> Using `Toolbar` as a collapsible group of controls is deprecated. Please use `ToolbarGroup` instead.

Should I migrate all the current `Toolbar` usages in the codebase or is there another approach to bypass the warnings on the tests?

diegohaz committed:
Update comments

gziolo commented:
> I added a deprecation warning to `Toolbar` and now many end-to-end tests are breaking because of the `console.warn`:
> 
> > Using `Toolbar` as a collapsible group of controls is deprecated. Please use `ToolbarGroup` instead.
> 
> Should I migrate all the current `Toolbar` usages in the codebase or is there another approach to bypass the warnings in the tests?

Yes, that's the point of having this check for warnings on the JS console. It ensures that we are always using recommended APIs and never use those which got deprecated.

gziolo commented on packages/components/src/toolbar/toolbar-container.js:
```diff
@@ -1,6 +1,22 @@
+
+	return (
+		<ToolbarContext.Provider value={ toolbarState }>
+			<Toolbar { ...toolbarState } aria-label={ accessibilityLabel } { ...props } />
```
`aria-label={ accessibilityLabel }` - we could build a helper which translates accessibility-related props to their ARIA version.

gziolo commented on packages/components/src/toolbar/test/index.js:
```diff
@@ -1,110 +0,0 @@
- */
-import Toolbar from '../';
-
-describe( 'Toolbar', () => {
```
Maybe we should keep them for backward compatibility with the note in the `describe` that it's a deprecated version?

In each test we would have to add:

```js
expect( console ).toHaveWarnedWith( 'Using `Toolbar` as a collapsible group of controls...' );
```

gziolo replied:
It’s probably fine to keep one of those tests only. I figured out you moved the existing tests to the suite which covers the toolbar group. It didn’t occur to me at first.

gziolo commented on packages/components/src/dropdown-menu/index.js:
```diff
@@ -30,6 +30,16 @@ function mergeProps( defaultProps = {}, props = {} ) {
 	return mergedProps;
 }
 
+function callAll( ...fns ) {
```
Is it really necessary? It seems like it would be easier to follow if those functions calls would be inlined like:

```js
onClick={
    ( ...args ) => {
        onToggle( ...args );
        mergedToggleProps. onClick( ...args );
    }
}
```

Even better, we could consider having this logic embedded into `mergeProps` helper. Feel free to update it if this would simplify things further.

gziolo commented:
Thanks also for sharing your feedback in
https://github.com/WordPress/gutenberg/pull/17875#issuecomment-543120002. 

> I think that's the path that will give us more flexibility to change things (even removing Reakit later, if needed) at the same time we provide more accessible ready-to-use components faster. But, of course, I'm always open to new ideas. :)

100% like what you wrote. It all sounds like a very good approach. We can discuss the next step once we have this PR sorted out. It looks very close to being ready. I'll give it a spin early next week. Let me know when it's ready for testing - I will apply `Needs Accessibility Feedback` label. You can also share it in the WordPress Slack #accessibility channel.

gziolo commented on packages/components/src/toolbar/index.js:
```diff
@@ -2,86 +2,39 @@
+export const __unstableToolbarContext = ToolbarContext;
 
-	if ( isCollapsed ) {
+function Toolbar( { className, accessibilityLabel, ...otherProps } ) {
```
This is missing now JSDOC comment. It should be added later.

etoledom commented:
Hey @diegohaz !

I did a smoke test on native mobile, and it looks great 👍 
At first it looked like there were a few regressions, but after updating this branch with current master everything was perfect.

diegohaz committed:
Test Toolbar

diegohaz committed:
Replace `callAll` by inline merge on DropdownMenu

diegohaz committed:
Add hint to deprecation warning

diegohaz committed:
Add jsdocs to Toolbar

diegohaz committed:
Import toolbar-group style into style.css

diegohaz committed:
Rename Toolbar/ToolbarGroup class names

Kept `.components-toolbar` for ToolbarGroup, so references to this selector don't break.

diegohaz committed:
Update Toolbar usage in components

diegohaz committed:
Update snapshots

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Update experimental Toolbar usage

diegohaz committed:
Update header toolbar

diegohaz committed:
Fix Fill not updating when Slot fillProps change

diegohaz committed:
Update Toolbar

diegohaz committed:
Move data-toolbar-button to the accessible toolbar button

diegohaz committed:
Update Toolbar stories

diegohaz committed:
Update BlockControls and BlockFormatControls slots

diegohaz committed:
Use ToolbarButton instead of IconButton?

Not sure about this

diegohaz commented:
Updated the code and did a lot of migrations. But it doesn't look good to me. The migration path is huge, specially because `Toolbar` now requires all toolbar items to be `ToolbarButton`, whereas most of Gutenberg is using `IconButton` instead. Some blocks are also using buttons with text, which `ToolbarButton` currently doesn't support.

I'm considering to accept `Button` instead (so both `IconButton` and `ToolbarButton` would be contemplated).

diegohaz committed:
Revert "Use ToolbarButton instead of IconButton?"

This reverts commit 5a03322508cacab6d02956c6567ddcfc9e8b24ad.

diegohaz committed:
Update styles

diegohaz committed:
Extend toolbar context to accept any button

diegohaz committed:
Separate toolbar context

diegohaz committed:
Update Toolbar Group styles

diegohaz committed:
Revert DropdownMenu to master version

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

gziolo commented:
@youknowriad - I would appreciate your feedback on this PR. I'm going to do an in-depth review and testing next week.

youknowriad commented:
So If I sum up my understanding here:

- the new `Toolbar` is basically the current `NavigatbleToolbar` that relies on the presence of `ToolbarButton`  components instead of using DOM to find the tabbable elements.

- the existing `Toolbar` is still there for BC concerns.

--- 
My main concerns is that we spent a lot of time thinkering about the block extensibility APIs and I see that this impacts how block authors should write their blocks. So I'd like a summary (before/after) in order to give a more in-depth feedback.

I do think that if the only difference is that we enforce `ToolbarButton`  instead of relying on DOM, it seems that maybe it's not worth the overhead especially since these block toolbars can be shared across different contexts, coming from slots, generated by hooks (see `useColors` hook)... and enforcing that a component called `ToolbarButton` is mounted/unmounted as a  children of a `Toolbar` could be very hard to achieve in the context of the block toolbar.

diegohaz committed:
Check if button is descendant of Toolbar

diegohaz committed:
Fix toolbar button

diegohaz committed:
Try some weird stuff

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Tweaks

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Update package-lock.json

diegohaz committed:
Migrate missing Toolbar

diegohaz committed:
Delay Toolbar focus on mount

diegohaz committed:
Fix Toolbar styling

diegohaz committed:
Fix tabThroughBlockToolbar e2e test

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Fix Tab navigation on block-hierarchy navigation test

diegohaz committed:
Try without isMounted on Slot

diegohaz committed:
Try without useMountedToolbarState

diegohaz commented:
The PR is finally ready for a full review. I updated the description to address the recent changes and @youknowriad comments.

> the new `Toolbar` is basically the current `NavigatbleToolbar` that relies on the presence of `ToolbarButton` components instead of using DOM to find the tabbable elements.

`NavigableToolbar` has some accessibility issues discussed on #15331. Besides fixing those issues, this PR creates a `Toolbar` component that is generic enough to be part of `@wordpress/components`. The API must be improved, but that can be done on next steps.

I changed the requirement from `ToolbarButton` to `Button`. `ToolbarButton` is still very limited. It's basically only an `IconButton`, and most of the toolbar items on Gutenberg doesn't use it.

These are the changes from the block author perspective:
```jsx
import {
  BlockControls,
  Button,
  IconButton,
  ToolbarButton,
  DropdownMenu,
} from '@wordpress/components';

// still works, but will fallback to the old navigable menu
// less accessible, but without breaking changes
<BlockControls>
  <button />
  <button />
  <button />
  <button />
</BlockControls>

// after
<BlockControls>
  <Button />
  <IconButton />
  <ToolbarButton />
  <DropdownMenu />
</BlockControls>
```

diegohaz commented:
Also, if this PR is too much (it feels like that for me), I'm open to make all the new features `__experimental`, remove the deprecation warnings for now and open other PRs implementing it on different parts of Gutenberg before marking it as stable.

gziolo commented on packages/components/src/index.native.js:
```diff
@@ -6,6 +6,8 @@ export { default as Dashicon } from './dashicon';
 export { default as Toolbar } from './toolbar';
 export { default as ToolbarButton } from './toolbar-button';
+export { default as ToolbarGroup } from './toolbar-group';
+export { createToolbarSlotFill, ToolbarSlot, ToolbarFill } from './toolbar-slot-fill';
```
Is it necessary to expose all those SlotFill related components in the public API?

diegohaz replied:
Ideally not. How is the approach to expose things to Gutenberg only?

gziolo replied:
Upon more detailed inspection, I see that only `createToolbarSlotFill` is used in different packages. Isn't it an issue with SlotFill that it doesn't propagate the context properly? There is a PR from @youknowriad where he tries to fix the issue for the Inspector slot: https://github.com/WordPress/gutenberg/pull/16807. I'm sure there is more as I remember that there was PR were another Slot was refactored to bubble virtually and this allowed to fix a similar issue. In fact, it turns out it was targeting the block toolbar, see #16421. I'm sure that Riad will have some good insights on what would be the best way to address it.

gziolo replied:
The related issue which is probably your use case as well: https://github.com/WordPress/gutenberg/issues/9203.

diegohaz replied:
The `Slot`/`Fill` module was the trickiest part of this PR for me. Indeed, I agree that those issues should be solved in the module itself. From what I understood (please, correct me if I'm wrong), this is how it currently works:
```jsx
const SlotContext = React.createContext();
const FillContext = React.createContext();

<SlotContext.Provider>
  <Slot name="a" />
  <Slot name="b" bubblesVirtually />
</SlotContext.Provider>

<FillContext.Provider>
  {/* will have access to SlotContext, but not FillContext, because it's rendered in the Slot React tree  */}
  <Fill name="a" />
  {/* will have access to FillContext, but not SlotContext, because it's rendered in this tree and portaled to the slot DOM tree */}
  <Fill name="b" />
</FillContext.Provider>
```

I understand that `bubblesVirtually` was added so `Fill`s could access contexts  in the same tree as custom blocks `edit`, right? Since `BlockControls` uses `bubblesVirtually`, it wouldn't have access to the `ToolbarContext`.

diegohaz replied:
The solution I found was passing the context state to the `Fill` tree through the `fillProps` so `Fill` would have access to both `FillContext` and `SlotContext` (`ToolbarContext` in this case).

There may be some way to solve that directly on the slot-fill module, but I'm afraid that it becomes even more frankenstein.

diegohaz replied:
Now that I'm thinking about it, putting the ToolbarContext higher in the block component tree would solve this without all the `ToolbarSlot`/`ToolbarFill` components. 😅

gziolo replied:
The principle we try to follow is, if not necessary, let's not add new public APIs because of backward-compatibility reasons. You need to maintain them for years in WordPress world :)

diegohaz replied:
I removed `ToolbarSlot`/`ToolbarFill` from `@wordpress/components`. Had to expose `ToolbarContext` though.

diegohaz committed:
Remove ToolbarSlotFill

diegohaz committed:
Put ToolbarProvider closer to the block contextual toolbar and the block edit

diegohaz committed:
Update BlockListBlock format

diegohaz committed:
Use Header Toolbar Provider when contextual toolbar is not shown

diegohaz committed:
Remove Contextual Toolbar context only when there is a fixed toolbar

diegohaz committed:
Revert to a1328b2

diegohaz committed:
Remove ToolbarSlotFill from components package

gziolo commented:
This patch no longer applies cleanly to the `master` branch. It needs to be rebased. I sorted it out by temporarily removing a few changes to be able to test the behavior in Storybook.

gziolo commented:
I did some extensive testing in Storybook and I discovered an issue with dropdown menus:

![toolbar](https://user-images.githubusercontent.com/699132/68472143-23241e00-0220-11ea-89c7-1fa25747a02c.gif)

Otherwise, all stories works exactly as expected ❤️

gziolo commented:
The best part is that everything works properly inside Gutenberg:

![toolbar-gutenberg](https://user-images.githubusercontent.com/699132/68473051-3f28bf00-0222-11ea-9156-5c0a6cf3dfd9.gif)

Arrow keys navigation works perfectly fine inside the toolbar.
Tabbing moves you to the next/previous focusable item outside of the toolbar. Tabbing back moves focus to the last active element.
Navigation between the toolbar and the dropdown is very intuitive and follows recommendations.

Amazing work @diegohaz 😍 

I also could see the deprecation messages displayed on the JS console but only because I couldn't apply the patch cleanly as mentioned in one of my previous comments. This is good of course.

![Screen Shot 2019-11-08 at 12 26 29](https://user-images.githubusercontent.com/699132/68473346-06d5b080-0223-11ea-8e2b-ce3a9568f571.png)

I tested different contexts of the toolbar placement and it works properly as well:
![toolbar-gutenberg-contexts](https://user-images.githubusercontent.com/699132/68473527-82376200-0223-11ea-98c2-1a8a8cfcb9af.gif)


There are some known issues:
-  with the dropdown menu implementation - tabbing should work differently
- with the order of visual elements which are focusable on smaller screens
However, this is something which is reported and totally out of the scope for this PR 😄

gziolo commented on packages/components/src/toolbar/index.js:
```diff
@@ -2,86 +2,51 @@
+ * To add controls, simply pass `ToolbarButton` components as children.
  *
- * @return {WPComponent} The rendered component.
+ * @param {Object} props
```
Param descriptions are missing in JSDoc comment.

gziolo commented on packages/components/src/toolbar/index.js:
```diff
@@ -2,86 +2,51 @@
+			<ToolbarContainer
+				ref={ ref }
+				// `ToolbarGroup` already uses components-toolbar for compatibility reasons
+				className={ classnames( 'components-accessible-toolbar', className ) }
```
I'm wondering whether we could keep the old class-name `components-toolbar` and add some modifiers to tweak the styles for the legacy use case.

diegohaz replied:
Can you explain better how it could be done?

My only concern is that the `.components-toolbar` selector is being used in other parts of Gutenberg, so I guess that other projects might also be using it as it were part of the public API.

gziolo replied:
@jasmussen - can you help to figure out how to perform the migration for class names?

This is what happens with this PR:

1. `Toolbar` becomes `ToolbarGroup`.
2. `Toolbar` should be now the wrapper for the collection of individual toolbar buttons and/or toolbar groups.
3. @diegohaz proposed we use a `components-accessible-toolbar` class name with a very bare set of styles for the rebranded `Toolbar` component.
4. In effect, all previous occurrences fo styles need to cover both `components-toolbar` and `components-toolbar-group` class names.
5. My question is, can you think of the way, where we could repurpose the old `.components-toolbar` to use instead of `.components-accessible-toolbar`?

To better picture it, can we use a trick like:

```scss
.components-toolbar {
    // some styles for the new Toolbar
   &.legacy {
       // some old styles which get applied when using legacy version
   }
}
```

mtias replied:
We should avoid setting the expectation that classes are public APIs.

gziolo reviewed (commented):
I left some more detailed feedback for the actual implementation as well.

gziolo commented on packages/block-editor/src/components/navigable-toolbar/index.js:
```diff
@@ -28,42 +33,69 @@ class NavigableToolbar extends Component {
-	}
+		// Toolbar items added via Portal (Slot bubblesVirtually) aren't added
+		// to the DOM right away
+		window.requestAnimationFrame( () => {
```
@youknowriad  and @aduth - do you know if we can avoid it?

aduth replied:
It's not ideal that they're not made available immediately. I'm unsure if the implementation of Slot/Fill can be changed to support what's need here.

I've not yet looked at the rest of the code proposed here, but one option to consider is if it might be possible to apply a `componentDidMount` to the elements _within_ the fill, to use as an indicator of when they're actually available in the DOM. I expect it could make things more challenging to coordinate (multiple fill elements, coordinating with an ancestor component).

diegohaz commented on packages/block-editor/src/components/block-controls/index.js:
```diff
@@ -10,15 +11,26 @@ import { ifBlockEditSelected } from '../block-edit/context';
-	</Fill>
-);
+function BlockControlsSlot( props ) {
+	const accessibleToolbarState = useContext( ToolbarContext );
```
`useToolbarContext` would be simply `() => useContext( ToolbarContext )`, right? We can do that, but nothing really changes. We would have also to expose a `ToolbarProvider` for the slotfill case. They can be `__unstable` though.

gziolo replied:
In that case, let's prefix it as experimental and revisit later:
```js
__ExperimentalToolbarContext
```

We use this technique to give it some time new public APIs to settle down before we decide to expose them and have to support for years.

diegohaz commented on packages/components/package.json:
```diff
@@ -42,6 +42,7 @@
 		"re-resizable": "^6.0.0",
 		"react-dates": "^17.1.1",
 		"react-spring": "^8.0.20",
+		"reakit": "^1.0.0-beta.10",
```
It's pretty much stable. [There's a few details](https://github.com/reakit/reakit/issues/429) I'd like to address before that. But probably we're being overly cautious there.

gziolo replied:
It sounds great, so it looks like the only thing to fix is the performance of menus which isn't a concern in this context 👍

diegohaz commented on packages/components/src/button/index.js:
```diff
@@ -41,13 +46,15 @@ export function Button( props, ref ) {
-		ref,
-	} );
+	// If it's within a toolbar, render with toolbar context instead.
+	if ( isWithinToolbar ) {
```
Yes! But also because it's impossible to provide a good migration path without greatly improving `ToolbarButton`. Ideally, `ToolbarButton` would be flexible enough for all use cases. Even on Gutenberg we use a lot of different types of buttons in the Toolbar. [I even tried to migrate them all to `ToolbarButton`](https://github.com/WordPress/gutenberg/pull/17875/commits/5a03322508cacab6d02956c6567ddcfc9e8b24ad), but right not it's not possible, so I reverted it and put the code on `Button` instead.

gziolo replied:
Have you considered using:
```js
element.closest( 'role="toolbar"' );
```
 instead?
I don't know if it is any better, but I used in the dropdown implementation to detect whether the focus is transferred to the modal dialog with this check
```js
document.activeElement.closest( '[role="dialog"]' )
```

diegohaz replied:
I'm using `.closest` inside `useIsWithinToolbar`.

diegohaz commented on packages/components/src/slot-fill/context.js:
```diff
@@ -60,6 +60,9 @@ class SlotFillProvider extends Component {
 			previousSlot.forceUpdate();
+
+			// See 'should re-render Fill when fillProps are updated' test.
+			this.forceUpdateFills( name );
```
The issue is that `Fill` wasn't being re-rendered when `fillProps` changed because [`contextValue`](https://github.com/WordPress/gutenberg/blob/c721a4d1daf3176f6118b9261d2821a9322679a7/packages/components/src/slot-fill/context.js#L37-L45) never changes (and therefore there's no need to re-render consumers), even when there are new `fillProps`. In fact, `fillProps` is not included in the context state. [It's accessed using the `Slot` instance](https://github.com/WordPress/gutenberg/blob/c721a4d1daf3176f6118b9261d2821a9322679a7/packages/components/src/slot-fill/fill.js#L58) that is returned by `context.getSlot()`. Both the getter function and the instance reference never change.

My first attempt was transforming `contextValue` into `this.state`, put `fillProps` there and update it when new `fillProps` arrived, which would make consumers re-render. Not sure why it hasn't been done that way in the first place. I'll be glad to refactor `Slot/Fill` in a separate PR as well.

gziolo replied:
Yes, it seems like its own PR would make it easier to review, test and land. It should be quite straightforward to tackle.

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Fix toolbar group styling

diegohaz committed:
Remove subscript prop from ToolbarGroup tests

diegohaz committed:
Fix DropdownMenu items don't receiving focus on Storybook

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

gziolo commented on packages/components/src/button/toolbar-button.js:
```diff
@@ -0,0 +1,33 @@
+		// tree and then decide whether to use the accessible Toolbar (which only
+		// accepts `ToolbarButton` as toolbar items) or fallback to the legacy
+		// `NavigableToolbar`, which accepts any tabbable element.
+		'data-toolbar-button': true,
```
It would be better if we could mark legacy focusable elements with the data attribute but I guess this is way much easier. I also can't think myself of any better way of doing it.

diegohaz replied:
The only way I can think is using [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver). That's also what I would try to use to make `NavigableContainer` add `tabindex="-1"` to inactive items.

gziolo commented:
Can you summarize in the comment what's left so we could decide if there any remaining blockers for this PR?

gziolo commented:
I did another round of testing and all my issues reported was fixed. UX wise, this PR is good to go in my opinion. Let's wait for the list of tasks left if any and decide how to proceed with the merge.

gziolo commented:
I see this when setting the full width for the Table block when in the mobile viewport, but this might be present in `master`:
![Screen Shot 2019-11-12 at 10 25 06](https://user-images.githubusercontent.com/699132/68658976-bae27e80-0536-11ea-97fe-c6b849776b0f.png)
Let me confirm. I will open a follow-up issue if that is the case.

diegohaz committed:
ToolbarContext -> __experimentalToolbarContext

diegohaz committed:
Add missing JSDocs to ToolbarGroup

diegohaz committed:
Fix ref merging on useIsWithinToolbar

diegohaz committed:
Add missing JSDoc descriptions

diegohaz commented:
> I see this when setting the full width for the Table block when in the mobile viewport, but this might be present in `master`

Checked it and it was introduced by this PR. Working on a fix.

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz commented:
> Can you summarize in the comment what's left so we could decide if there any remaining blockers for this PR?

Marked the comments that were addressed as `resolved`. The ones missing are generally discussions.

Future tasks I can think of:

- Improve `ToolbarButton` so it can be used for more cases
- Refactor `SlotFill`

diegohaz committed:
Update stories

jasmussen commented:
> I see this when setting the full width for the Table block when in the mobile viewport, but this might be present in master:

Hey, Diego is right, https://github.com/WordPress/gutenberg/pull/17877 is responsible for this. I created #18469 to fix it.

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

ItsJonQ reviewed (approved):
@diegohaz Amazing job with this PR!

I tested this locally in Gutenberg with keyboard + voiceover. It works as expected. There were a couple of hiccups, but those interactions already exist in Gutenberg today (unfortunately. We'll make it better!).

Based on the convo with @gziolo , I think a great fast follow PR would be to improve the `SlotFIll` related forceUpdate work around.

Thank you so much for all of your hard work.
Double thank you for your efforts and advocacy for Reakit and a11y!

Very excited to see how we can systematically improve the experience of Gutenberg as we start refactoring with Reakit within the components.

❤️

youknowriad commented:
It's not clear to me yet how this impacts our existing public APIs and what's the potential of breakage of existing usage. Anyone able to write a summary?

diegohaz commented:
@youknowriad I wrote a summary here: https://github.com/WordPress/gutenberg/pull/17875#issuecomment-549202734

It shouldn't break the current usage unless they're doing things within `<BlockControls>` that diverge a lot from what Gutenberg does with built-in blocks.

youknowriad commented:
To be completely honest, I don't feel confident merging this PR as is. My concerns are:

1- It is too big, I wonder if we can split it to increase confidence in the changes we're making.
2- Things like that:

> It shouldn't break the current usage unless they're doing things within <BlockControls> that diverge a lot from what Gutenberg does with built-in blocks.

This statement is a little bit unprecise to me. We can expect that plugin authors do all kind of weird things and Backwards compatibility is really important for WordPress, we can't break existing APIs.

Some exceptions are possible when we know exactly what we're breaking, how we're breaking it and if the breakage is small enough. In that case, we can write a dev note for plugin authors to let them know how they should update their code to avoid breakage.

diegohaz commented:
@youknowriad 
I think this PR could be simplified if all the new features are made experimental and no deprecation is made. Gutenberg itself wouldn't change (it would still use the old Toolbar without deprecation warnings) and the accessible Toolbar adoption could be broken into more PRs.

In the meanwhile, other projects using `@wordpress/components` could still benefit from this (knowing that it's experimental) and we would have room to try and change things.

youknowriad commented:
> I think this PR could be simplified if all the new features are made experimental and no deprecation is made. Gutenberg itself wouldn't change (it would still use the old Toolbar without deprecation warnings) and the accessible Toolbar adoption could be broken into more PRs.

If we make it experimental, when do you think we could get out of the experimental phase? What concerns me the most is not the deprecations. Deprecations are good because they let developpers know what they should do. 

It seems like you're saying that we can add the new components without touching the existing Core blocks. That is reassuring. So aside deprecations which can potentially break e2e tests, if I'm a block author and I don't touch my code, is there something that can break?

diegohaz commented:
> If we make it experimental, when do you think we could get out of the experimental phase? What concerns me the most is not the deprecations. Deprecations are good because they let developers know what they should do.

How is the process that is currently used to take things out of the experimental phase?

The removal of the deprecation would allow us to change between both implementations and even revert this one entirely if it doesn't work out. One thing that I learned while working on this PR is that making things `__experimental` and not deprecating stuff in the first step is the best way to introduce new features on Gutenberg (as long as you continue working on that).

Steps I would propose (they could be separate PRs):

1. Simplify this PR so only `packages/components/src/toolbar-*` is touched. 

    The accessible toolbar would be enabled by passing an `__experimentalAccessibilityLabel` prop to `Toolbar`. Only `ToolbarButton` would be accepted as a toolbar item (with the current state of this PR, it's `Button`, but only because `ToolbarButton` is still not flexible enough to serve all use cases).
  
    At this point, other projects using `@wordpress/components` could start experimenting it for simple toolbars.

2. Improve `ToolbarButton` so that it serves more use cases than just being an `IconButton`.

    At this point, the Gutenberg header toolbar (without the fixed block toolbar) could use it.

3. Refactor `SlotFill` so that the workarounds introduced by the current state of this PR (forcing fills to re-render when `fillProps` change, for example) are not necessary.

    At this point, block toolbars could start using it.

4. Convert the fixed toolbar buttons (block switcher, more tools etc.) to `ToolbarButton`.

    At this point, custom blocks could start using it, as long as they use `ToolbarButton` inside `BlockControls` or `BlockFormatControls`.

5. Convert core block toolbars.

    We could convert one per PR (for example, start with the Paragraph toolbar).

6. Let it be for a while.

7. If it proves to be useful and works out, deprecate the old usage and remove the `__experimental` prefix. Otherwise, revert it altogether, which would break the code for people using `__experimentalAccessibilityLabel` explicitly, but custom blocks would seamlessly rollback to the old usage.

youknowriad commented:
An iterative process like that sounds good to me 👍

diegohaz commented:
Closing in favor of #18534
